File: lib/domaindive/analysis.ex
defmodule Domaindive.Analysis do
  @moduledoc """
  Context for domain analysis operations.
  """

  import Ecto.Query, warn: false
  alias Domaindive.Repo
  alias Domaindive.Analysis.DomainAnalysis
  alias Domaindive.Analysis.WhoisService
  alias Domaindive.Analysis.DnsService
  alias Domaindive.Analysis.SslService
  alias Domaindive.Analysis.HttpService
  alias Domaindive.Analysis.GeolocationService

  @cache_ttl_minutes 5

  @doc """
  Gets or creates a domain analysis for the given address.

  If a cached analysis exists and is older than #{@cache_ttl_minutes} minutes, it will be refreshed.
  Returns `{:ok, domain_analysis, :cached}` for cached data or `{:ok, domain_analysis, :fresh}` for new/refreshed data.
  """
  def get_or_create_analysis(address, force_refresh \\ false) do
    normalized_address = normalize_address(address)

    case Repo.get_by(DomainAnalysis, address: normalized_address) do
      nil ->
        case create_new_analysis(normalized_address) do
          {:ok, analysis} -> {:ok, analysis, :fresh}
          error -> error
        end

      existing_analysis ->
        if force_refresh || needs_refresh?(existing_analysis) do
          case refresh_analysis(existing_analysis) do
            {:ok, analysis} -> {:ok, analysis, :fresh}
            _error -> {:ok, existing_analysis, :cached}
          end
        else
          {:ok, existing_analysis, :cached}
        end
    end
  end

  @doc """
  Checks if an analysis needs to be refreshed based on its updated_at timestamp.
  """
  def needs_refresh?(%DomainAnalysis{updated_at: nil}), do: true

  def needs_refresh?(%DomainAnalysis{updated_at: updated_at}) do
    updated_at_utc = DateTime.from_naive!(updated_at, "Etc/UTC")
    age_minutes = DateTime.diff(DateTime.utc_now(), updated_at_utc, :minute)
    age_minutes >= @cache_ttl_minutes
  end

  @doc """
  Calculates seconds until the next refresh is needed.
  """
  def seconds_until_refresh(%DomainAnalysis{updated_at: nil}), do: 0

  def seconds_until_refresh(%DomainAnalysis{updated_at: updated_at}) do
    updated_at_utc = DateTime.from_naive!(updated_at, "Etc/UTC")
    refresh_time = DateTime.add(updated_at_utc, @cache_ttl_minutes * 60, :second)
    max(0, DateTime.diff(refresh_time, DateTime.utc_now(), :second))
  end

  defp refresh_analysis(existing_analysis) do
    perform_analysis(existing_analysis.address)
    |> case do
      {:ok, attrs} ->
        existing_analysis
        |> DomainAnalysis.changeset(attrs)
        |> Repo.update()

      error ->
        error
    end
  end

  defp create_new_analysis(normalized_address) do
    case perform_analysis(normalized_address) do
      {:ok, attrs} ->
        %DomainAnalysis{}
        |> DomainAnalysis.changeset(attrs)
        |> Repo.insert()

      error ->
        error
    end
  end

  defp perform_analysis(normalized_address) do
    whois_data =
      case WhoisService.fetch(normalized_address) do
        {:ok, data} -> data
        {:error, _} -> nil
      end

    dns_records =
      try do
        dns_data = DnsService.fetch_records(normalized_address)
        Jason.encode!(dns_data)
      rescue
        e ->
          IO.inspect(e, label: "DNS encoding error")
          Jason.encode!(%{})
      end

    nameservers =
      case DnsService.fetch_nameservers(normalized_address) do
        {:ok, data} -> Jason.encode!(data)
        {:error, _} -> Jason.encode!([])
      end

    {:ok, ssl_data} = SslService.fetch_certificate(normalized_address)
    ssl_info = Jason.encode!(ssl_data)

    http_response =
      case HttpService.fetch_headers(normalized_address) do
        {:ok, data} -> Jason.encode!(data)
        {:error, _} -> Jason.encode!(%{})
      end

    geolocation =
      case get_geolocation_for_domain(normalized_address) do
        {:ok, data} -> Jason.encode!(data)
        {:error, _} -> Jason.encode!(%{})
      end

    {:ok,
     %{
       address: normalized_address,
       whois_data: whois_data,
       dns_records: dns_records,
       nameservers: nameservers,
       ssl_info: ssl_info,
       http_response: http_response,
       geolocation: geolocation
     }}
  end

  defp get_geolocation_for_domain(domain) do
    try do
      case DnsService.fetch_records(domain) do
        # Match when we have at least one A record with a value
        %{a: [%{value: ip} | _]} ->
          GeolocationService.fetch_location(ip)

        # No A records found or empty list
        _ ->
          {:error, :no_ip}
      end
    rescue
      e ->
        # Log the actual error for debugging
        IO.inspect(e, label: "Geolocation lookup error")
        {:error, :lookup_failed}
    end
  end

  @doc """
  Gets a domain analysis by address.

  Returns `nil` if no analysis exists.

  ## Examples

      iex> get_analysis_by_address("example.com")
      %DomainAnalysis{}

      iex> get_analysis_by_address("nonexistent.com")
      nil

  """
  def get_analysis_by_address(address) do
    normalized_address = normalize_address(address)
    Repo.get_by(DomainAnalysis, address: normalized_address)
  end

  defp normalize_address(address) do
    address
    |> String.trim()
    |> String.downcase()
    |> String.replace(~r/^https?:\/\//, "")
    |> String.replace(~r/\/$/, "")
  end
end

File: src/lib/server/analysis.ts
import db, { type DomainAnalysis } from "./db.js";
import { fetchDnsRecords, fetchNameservers } from "./dns.js";
import { fetchGeolocation } from "./geolocation.js";
import { fetchHttpHeaders } from "./http.js";
import { fetchSslCertificate } from "./ssl.js";
import { fetchWhois } from "./whois.js";

const CACHE_TTL_MINUTES = 5;

function normalizeAddress(address: string): string {
    return address
        .trim()
        .toLowerCase()
        .replace(/^https?:\/\//, "")
        .replace(/\/$/, "");
}

function needsRefresh(updatedAt: string | null): boolean {
    if (!updatedAt) return true;
    const now = new Date();
    const updated = new Date(updatedAt);
    const ageMinutes = (now.getTime() - updated.getTime()) / 1000 / 60;
    return ageMinutes >= CACHE_TTL_MINUTES;
}

function validateDomain(address: string): boolean {
    const domainRegex = /^[a-zA-Z0-9][a-zA-Z0-9-_.]+\.[a-zA-Z]{2,}$/;
    return domainRegex.test(address);
}

async function performAnalysis(address: string) {
    let whoisData: string | null = null;
    try {
        whoisData = await fetchWhois(address);
    } catch (e) {
        console.error(e);
    }

    const dnsRecords = await fetchDnsRecords(address);
    const nameservers = await fetchNameservers(address);
    const sslInfo = await fetchSslCertificate(address);
    const httpResponse = await fetchHttpHeaders(address);

    let geolocation = {};
    if (dnsRecords.a && dnsRecords.a.length > 0) {
        const ip = dnsRecords.a[0].value;
        geolocation = await fetchGeolocation(ip);
    }

    return {
        address,
        whois_data: whoisData,
        dns_records: JSON.stringify(dnsRecords),
        nameservers: JSON.stringify(nameservers),
        ssl_info: JSON.stringify(sslInfo),
        http_response: JSON.stringify(httpResponse),
        geolocation: JSON.stringify(geolocation),
    };
}

export async function getOrCreateAnalysis(address: string, forceRefresh = false) {
    const normalizedAddress = normalizeAddress(address);

    if (!validateDomain(normalizedAddress)) {
        throw new Error("must be a valid domain name with a top-level domain (e.g., example.com)");
    }

    const query = db.query<DomainAnalysis, string>("SELECT * FROM domain_analyses WHERE address = ?");
    const existing = query.get(normalizedAddress);

    if (existing && !forceRefresh && !needsRefresh(existing.updated_at)) {
        return { analysis: existing, cacheStatus: "cached" as const };
    }

    const analysisData = await performAnalysis(normalizedAddress);

    if (existing) {
        const updateQuery = db.query(`
			UPDATE domain_analyses
			SET whois_data = $whois_data, dns_records = $dns_records, nameservers = $nameservers,
				ssl_info = $ssl_info, http_response = $http_response, geolocation = $geolocation,
				updated_at = CURRENT_TIMESTAMP
			WHERE address = $address
		`);
        updateQuery.run({
            $whois_data: analysisData.whois_data,
            $dns_records: analysisData.dns_records,
            $nameservers: analysisData.nameservers,
            $ssl_info: analysisData.ssl_info,
            $http_response: analysisData.http_response,
            $geolocation: analysisData.geolocation,
            $address: normalizedAddress,
        });
        const updated = query.get(normalizedAddress) as DomainAnalysis;
        return { analysis: updated, cacheStatus: "fresh" as const };
    } else {
        const insertQuery = db.query(`
			INSERT INTO domain_analyses
			(address, whois_data, dns_records, nameservers, ssl_info, http_response, geolocation)
			VALUES ($address, $whois_data, $dns_records, $nameservers, $ssl_info, $http_response, $geolocation)
		`);
        const result = insertQuery.run({
            $address: analysisData.address,
            $whois_data: analysisData.whois_data,
            $dns_records: analysisData.dns_records,
            $nameservers: analysisData.nameservers,
            $ssl_info: analysisData.ssl_info,
            $http_response: analysisData.http_response,
            $geolocation: analysisData.geolocation,
        });
        const getByIdQuery = db.query<DomainAnalysis, bigint>("SELECT * FROM domain_analyses WHERE id = ?");
        const created = getByIdQuery.get(BigInt(result.lastInsertRowid)) as DomainAnalysis;
        return { analysis: created, cacheStatus: "fresh" as const };
    }
}

export function secondsUntilRefresh(updatedAt: string | null): number {
    if (!updatedAt) return 0;
    const now = new Date();
    const updated = new Date(updatedAt);
    const refreshTime = new Date(updated.getTime() + CACHE_TTL_MINUTES * 60 * 1000);
    return Math.max(0, Math.floor((refreshTime.getTime() - now.getTime()) / 1000));
}
